torch/csrc/jit/codegen/cuda/ir_internal_nodes.h

```diff
@@ -267,5 +267,168 @@ struct TORCH_CUDA_API Reorder : public Expr {
   const std::vector<int> pos2axis_;
 };
 
+
+/*
+ * ForLoop provides scoping around an int iterator from 0 to range. Exprs placed
+ * in its body are considered inside the scope of the for loop. In the future
+ * the implementation should look quite different so that we can do proper
+ * dependency annalysis like in Fusion.
+ *
+ * TODO: Change implmentation of Exprs contained in the scope to be more similar
+ * to Fusion where we can do proper dependency analysis.
+ */
+struct TORCH_API ForLoop : public Expr {
+  ~ForLoop() = default;
+  ForLoop(
+      Int* _index,
+      IterDomain* _range,
+      const std::vector<const Expr*>& _body);
+
+  ForLoop(const ForLoop& other) = delete;
+  ForLoop& operator=(const ForLoop& other) = delete;
+
+  ForLoop(ForLoop&& other) = delete;
+  ForLoop& operator=(ForLoop&& other) = delete;
+
+  Int* index() const noexcept {
+    return index_;
+  }
+  IterDomain* range() const noexcept {
+    return range_;
+  }
+
+  const std::vector<const Expr*>& body() const noexcept {
+    return body_;
+  }
+
+  void add_expr(const Expr* e) {
+    body_.push_back(e);
+  }
+
+  void remove_expr(const Expr* e);
+  bool sameAs(const ForLoop* other) const;
+
+ private:
+  Int* const index_;
+  IterDomain* const range_;
+  std::vector<const Expr*> body_;
+};
+
+
+/*
+ * IfThenElse provides scoping for an boolean operator. Exprs placed in its body
+ * are considered inside the scope of the if statement. In the future the
+ * implementation should look quite different so that we can do proper
+ * dependency annalysis like in Fusion.
+ *
+ * TODO: Change implmentation of Exprs contained in the scope to be more similar
+ * to Fusion where we can do proper dependency analysis.
+ */
+struct TORCH_API IfThenElse : public Expr {
+  ~IfThenElse() = default;
+  IfThenElse(
+      Val* _cond,
+      const std::vector<const Expr*>& _if_body,
+      const std::vector<const Expr*>& _else_body = {});
+
+  IfThenElse(const IfThenElse& other) = delete;
+  IfThenElse& operator=(const IfThenElse& other) = delete;
+
+  IfThenElse(IfThenElse&& other) = delete;
+  IfThenElse& operator=(IfThenElse&& other) = delete;
+
+  Val* cond() const noexcept {
+    return cond_;
+  }
+  const std::vector<const Expr*>& if_body() const noexcept {
+    return if_body_;
+  }
+  const std::vector<const Expr*>& else_body() const noexcept {
+    return else_body_;
+  }
+
+  void add_if_expr(const Expr* e) {
+    if_body_.push_back(e);
+  }
+  void add_else_expr(const Expr* e) {
+    else_body_.push_back(e);
+  }
+
+  bool hasElse() const noexcept {
+    return !else_body_.empty();
+  }
+
+  bool sameAs(const IfThenElse* other) const;
+  
+ private:
+  // TODO: Why is the pointer const and not what's in the object?
```

Comment:
Primarily because the const model needs rework, which is not easy to do. I've tried a couple times to improve it but it can be challenging. You can try to do this and we can see how far we can get, however if we hit a dead end, likely it's fine to back it out until we get far enough that we're convinced we're ready for a const-ness model refactor.